File: apps/composites/assign-job/src/services.py
import asyncio
import logging
import os

import httpx

from .publisher import publish_job_assigned

logger = logging.getLogger(__name__)


async def get_metrics(client: httpx.AsyncClient, contractor_id: str) -> dict:
  url = f"{os.getenv('METRICS_URL')}/api/metrics/{contractor_id}"
  resp = await client.get(url)
  resp.raise_for_status()

  rows = resp.json()["metrics"]
  return {
    "contractor_id": contractor_id,
    "total_jobs": len(rows),
    "total_score": sum(m["score_delta"] for m in rows),
  }


def select_best(contractor_data: list[dict]) -> dict:
  return sorted(contractor_data, key=lambda c: (c["total_jobs"], -c["total_score"]))[0]


async def _search_contractors(
  client: httpx.AsyncClient, sector_code: str, category_code: str
) -> list[dict]:
  url = f"{os.getenv('CONTRACTOR_API_URL')}/contractors/search"
  resp = await client.get(
    url, params={"SectorCode": sector_code, "CategoryCode": category_code}
  )
  resp.raise_for_status()
  return resp.json()


async def _fetch_metrics_for_contractors(
  client: httpx.AsyncClient, contractors: list[dict]
) -> list[dict]:
  return await asyncio.gather(
    *(get_metrics(client, c["ContractorUuid"]) for c in contractors)
  )


async def _create_assignment(
  client: httpx.AsyncClient, case_id: str, contractor_id: str
) -> dict:
  url = f"{os.getenv('ASSIGNMENT_URL')}/assignments"
  resp = await client.post(
    url,
    json={"case_id": case_id, "contractor_id": contractor_id, "source": "AUTO_ASSIGN"},
  )
  resp.raise_for_status()
  return resp.json()


async def assign_contractor(case_id: str, postal_code: str, category_code: str) -> dict:
  sector_code = postal_code[:2]

  async with httpx.AsyncClient() as client:
    contractors = await _search_contractors(client, sector_code, category_code)
    if not contractors:
      raise ValueError("No eligible contractors found")  # noqa: TRY003

    contractor_data = await _fetch_metrics_for_contractors(client, contractors)
    best = select_best(contractor_data)

    assignment = await _create_assignment(client, case_id, best["contractor_id"])

    try:
      await publish_job_assigned(
        assignment_id=assignment["id"],
        case_id=case_id,
        contractor_id=best["contractor_id"],
      )
    except Exception:
      logger.exception("Failed to publish Job_Assigned for case %s", case_id)
      raise

    return assignment

File: apps/composites/assign-job/src/services.ts
import type { AssignmentAtomType } from "@townops/assignment-atom";
import type { ContractorAtomType } from "@townops/contractor-atom";
import type { MetricsAtomType } from "@townops/metrics-atom";
import { logger, rabbitmqClient } from "@townops/shared-ts";
import { hc } from "hono/client";

import { env } from "./env";

export interface ContractorMetrics {
  contractorId: string;
  totalJobs: number;
  totalScore: number;
}

/**
 * Search for eligible contractors via the Contractor atom.
 */
export async function searchContractors(
  sectorCode: string,
  categoryCode: string
): Promise<
  { id: string; name: string; email: string; contactNum: string | null }[]
> {
  const client = hc<ContractorAtomType>(env.CONTRACTOR_ATOM_URL);
  const res = await client.api.contractors.search.$get({
    query: { sectorCode, categoryCode },
  });
  if (!res.ok) {
    throw new Error(`Contractor search failed: HTTP ${res.status}`);
  }
  const data = await res.json();
  return data.contractors as {
    id: string;
    name: string;
    email: string;
    contactNum: string | null;
  }[];
}

/**
 * Fetch full contractor details by ID from the Contractor atom.
 */
export async function getContractorById(id: string): Promise<{
  id: string;
  name: string;
  email: string;
  contactNum: string | null;
}> {
  const client = hc<ContractorAtomType>(env.CONTRACTOR_ATOM_URL);
  const res = await client.api.contractors[":id"].$get({
    param: { id },
  });
  if (!res.ok) {
    throw new Error(`Contractor fetch failed for ${id}: HTTP ${res.status}`);
  }
  const data = await res.json();
  return data.contractor as {
    id: string;
    name: string;
    email: string;
    contactNum: string | null;
  };
}

/**
 * Fetch aggregated metrics for a single contractor from the Metrics atom.
 */
export async function getContractorMetrics(
  contractorId: string
): Promise<ContractorMetrics> {
  const client = hc<MetricsAtomType>(env.METRICS_ATOM_URL);
  const res = await client.api.metrics[":contractor_id"].$get({
    param: { contractor_id: contractorId },
  });

  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (!res.ok) {
    throw new Error(
      `Metrics fetch failed for contractor ${contractorId}: HTTP ${res.status}`
    );
  }

  const data = await res.json();
  const rows = data.metrics;
  return {
    contractorId,
    totalJobs: rows.length,
    totalScore: rows.reduce(
      (sum: number, m: { scoreDelta: number }) => sum + m.scoreDelta,
      0
    ),
  };
}

/**
 * Select the best contractor: fewest jobs first, then highest total score.
 */
export function selectBest(
  contractors: ContractorMetrics[]
): ContractorMetrics {
  const best = [...contractors].toSorted(
    (a, b) => a.totalJobs - b.totalJobs || b.totalScore - a.totalScore
  )[0];
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (!best) {
    throw new Error("Cannot select best from empty contractor list");
  }
  return best;
}

/**
 * Create a new assignment record in the Assignment atom.
 */
export async function createAssignment(
  caseId: string,
  contractorId: string
): Promise<{ id: string; caseId: string; contractorId: string }> {
  const responseDueAt = new Date(Date.now() + 60 * 1000).toISOString();
  const client = hc<AssignmentAtomType>(env.ASSIGNMENT_ATOM_URL);
  const res = await client.api.assignments.$post({
    json: {
      caseId,
      contractorId,
      source: "AUTO_ASSIGN",
      responseDueAt,
    },
  });

  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (!res.ok) {
    throw new Error(`Assignment creation failed: HTTP ${res.status}`);
  }

  const data = await res.json();
  const assignment = data.assignments;
  return assignment as { id: string; caseId: string; contractorId: string };
}

/**
 * Full orchestration: search → rank → assign → publish.
 */
export async function assignContractor(
  caseId: string,
  postalCode: string,
  categoryCode: string
): Promise<void> {
  const sectorCode = postalCode.slice(0, 2);

  logger.info(
    { caseId, sectorCode, categoryCode },
    "assign-job: searching contractors"
  );

  const contractors = await searchContractors(sectorCode, categoryCode);
  if (contractors.length === 0) {
    throw new Error(
      `No eligible contractors found for sector=${sectorCode} category=${categoryCode}`
    );
  }

  const metricsResults = await Promise.all(
    contractors.map((c) => getContractorMetrics(c.id))
  );

  const best = selectBest(metricsResults);
  logger.info(
    { caseId, contractorId: best.contractorId },
    "assign-job: selected best contractor"
  );

  const [assignment, contractorDetail] = await Promise.all([
    createAssignment(caseId, best.contractorId),
    getContractorById(best.contractorId),
  ]);

  await rabbitmqClient.publishToQueue(
    "sla-timers-queue",
    {
      assignment_id: assignment.id,
      case_id: caseId,
      contractor_id: best.contractorId,
      postal_code: postalCode,
      category_code: categoryCode,
    },
    { expirationMs: 60_000 }
  );

  await rabbitmqClient.publish("townops.events", "job.assigned", {
    assignmentId: assignment.id,
    caseId,
    contractorId: best.contractorId,
    contractorName: contractorDetail.name,
    contractorEmail: contractorDetail.email,
    contractorContact: contractorDetail.contactNum,
    status: "PENDING_ACCEPTANCE",
    email: contractorDetail.email,
  });

  logger.info(
    { caseId, assignmentId: assignment.id },
    "assign-job: job.assigned published"
  );
}
